gather.go

```diff
@@ -71,45 +71,67 @@ func (a *Agent) GatherCandidates() error {
 
 func (a *Agent) gatherCandidates() {
 	gatherStateUpdated := make(chan bool)
-	if err := a.run(func(agent *Agent) {
-		a.gatheringState = GatheringStateGathering
-		close(gatherStateUpdated)
-	}); err != nil {
-		a.log.Warnf("failed to set gatheringState to GatheringStateGathering for gatherCandidates: %v", err)
-		return
-	}
-	<-gatherStateUpdated
-
-	for _, t := range a.candidateTypes {
-		switch t {
-		case CandidateTypeHost:
-			a.gatherCandidatesLocal(a.networkTypes)
-		case CandidateTypeServerReflexive:
-			a.gatherCandidatesSrflx(a.urls, a.networkTypes)
-			if a.extIPMapper != nil && a.extIPMapper.candidateType == CandidateTypeServerReflexive {
-				a.gatherCandidatesSrflxMapped(a.networkTypes)
+
+	a.chanCandidate = make(chan Candidate, 1)
+	var closeChanCandidateOnce sync.Once
+	go func() {
+		for c := range a.chanCandidate {
+			if onCandidateHdlr, ok := a.onCandidateHdlr.Load().(func(Candidate)); ok {
+				onCandidateHdlr(c)
 			}
-		case CandidateTypeRelay:
-			if err := a.gatherCandidatesRelay(a.urls); err != nil {
-				a.log.Errorf("Failed to gather relay candidates: %v\n", err)
+		}
+		if onCandidateHdlr, ok := a.onCandidateHdlr.Load().(func(Candidate)); ok {
+			onCandidateHdlr(nil)
+		}
+	}()
+
+	go func() {
+		defer func() {
+			closeChanCandidateOnce.Do(func() {
+				close(a.chanCandidate)
+			})
+		}()
+
+		if err := a.run(func(agent *Agent) {
+			a.gatheringState = GatheringStateGathering
+			close(gatherStateUpdated)
+		}); err != nil {
+			a.log.Warnf("failed to set gatheringState to GatheringStateGathering for gatherCandidates: %v", err)
+			return
+		}
+		<-gatherStateUpdated
+
+		for _, t := range a.candidateTypes {
+			switch t {
+			case CandidateTypeHost:
+				a.gatherCandidatesLocal(a.networkTypes)
+			case CandidateTypeServerReflexive:
+				a.gatherCandidatesSrflx(a.urls, a.networkTypes)
+				if a.extIPMapper != nil && a.extIPMapper.candidateType == CandidateTypeServerReflexive {
+					a.gatherCandidatesSrflxMapped(a.networkTypes)
+				}
+			case CandidateTypeRelay:
+				if err := a.gatherCandidatesRelay(a.urls); err != nil {
+					a.log.Errorf("Failed to gather relay candidates: %v\n", err)
+				}
 			}
 		}
-	}
-	if err := a.run(func(agent *Agent) {
-		if a.onCandidateHdlr != nil {
-			a.chanCandidateCallback <- func() { a.onCandidateHdlr(nil) }
+		if err := a.run(func(agent *Agent) {
```

Comment:
I would combine the `run` calls. This just exists to allow 'cancelling' of queued tasks when `Close()` is called on the agent. Since the tasks are similar I say squash them.